Snippets as extra arguments on Twister CLI leads to early CMake error

**Describe the bug**

Whenever you call Twister with the command line argument `--extra-args SNIPPET="<valid-snippet-name>"`, CMake aborts with the error message `zephyr_set(... SCOPE snippets) doesn't exist.`. This happens very early on in Zephyr bootstrapping, more specifically when including `cmake/modules/snippets.cmake`.

The analogous usage for a simple build, on the other hand, does not result in any error message, so both `west build -S <valid-snippet-name>` and `west build -- -DSNIPPET="<valid-snippet-name>"` work as described and documented. So far, I myself was not able to find the cause of the non-existent scope for snippets in the case of Twister and need help from experts.

**To Reproduce**

Here is a more fancy console log:

```
west twister --verbose --inline-logs --cmake-only \
     --extra-args SNIPPET="cdc-acm-console" \
     --platform nucleo_f767zi \
     --test tests/drivers/uart/uart_async_api/drivers.uart.async_api
```

…breaks with (CMake output in `build.log`):

```
-- Application: /.../workspace-zephyr/zephyr/tests/drivers/uart/uart_async_api
-- CMake version: 3.27.4
-- Found Python3: /.../workspace-zephyr/.env/bin/python (found suitable version "3.8.10", minimum required is "3.8") found components: Interpreter 
-- Cache files will be written to: /home/user/.cache/zephyr
-- Zephyr version: 3.4.99 (/.../workspace-zephyr/zephyr)
-- Found west (found suitable version "1.1.0", minimum required is "0.14.0")
-- Board: nucleo_f767zi
CMake Error at zephyr/cmake/modules/extensions.cmake:2764 (message):
  zephyr_set(...  SCOPE snippets) doesn't exists.
Call Stack (most recent call first):
  twister-out/nucleo_f767zi/tests/drivers/uart/uart_async_api/drivers.uart.async_api/zephyr/snippets_generated.cmake:25 (zephyr_set)
  zephyr/cmake/modules/snippets.cmake:94 (include)
  zephyr/cmake/modules/snippets.cmake:121 (zephyr_process_snippets)
  zephyr/cmake/modules/zephyr_default.cmake:129 (include)
  zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:75 (include)
  zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:92 (include_boilerplate)
  zephyr/cmake/package_helper.cmake:115 (find_package)
```

**Expected behavior**

The expectation here would be that Twister behaves the same way with snippets as it does when passing a shield stack via extra arguments. Both `--extra-args SNIPPET="<valid-snippet-name>"` and `--extra-args SHIELD="<valid-shield-name>"` should not lead to any CMake errors, alone or together.

**Impact**

Announcement for the community, but showstopper for [Bridle](https://github.com/tiacsys/bridle). Currently I'm trying to convert a [board-specific shield with CPU clock tweaks](https://github.com/tiacsys/bridle/tree/main/boards/shields/can_timing_adj) to a snippet so that the default [Zephyr CAN timing test](https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/can/timing) passes.

**Logs and console output**
See above.

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain Zephyr SDK 0.16.1
 - main branch, commit SHA 341590545ac80b7a2abd9902b7f10d577323ccde

**Additional context**
Non.

> Give #63232 a try

Wow, quick response. Many thanks. Looks like right direction. Twister now finds downstream snippets in modules, too. BUT: now we run into exactly the same error as descript on top of this bug report. Let me give you the sample session.

Twister called by west:
```
west twister --cmake-only \
             --alt-config-root bridle/tests/drivers/can/timing \
             --testsuite-root zephyr/tests/drivers/can/timing
```

Exit now with:
```
INFO    - Using Ninja..
INFO    - Zephyr version: zephyr-v3.4.0-4509-g5f7fee5b9139
INFO    - Using 'zephyr' toolchain.
INFO    - Using alternative configuration from /.../workspace/bridle/tests/drivers/can/timing/testcase.yaml
INFO    - Selecting default platforms per test case
INFO    - Building initial testsuite list...
INFO    - Writing JSON report /.../workspace/twister-out/testplan.json
INFO    - JOBS: 8
INFO    - Adding tasks to the queue...
INFO    - Added initial list of jobs to queue
ERROR   - Cmake build failure: /.../workspace/zephyr/tests/drivers/can/timing for tiac_magpie
ERROR   - tiac_magpie               tests/drivers/can/timing/drivers.can.timing         ERROR : Cmake build failure
ERROR   - see: /.../workspace/twister-out/tiac_magpie/tests/drivers/can/timing/drivers.can.timing/build.log
… … …
```

Build log is (`.../tests/drivers/can/timing/drivers.can.timing/build.log`):
```
Loading Zephyr module(s) (Zephyr base): zephyr_default:dts
-- Application: /.../workspace/zephyr/tests/drivers/can/timing
-- CMake version: 3.27.4
Loading Bridle default modules (Freestanding).
-- Bridle version: 3.4.99 (/.../workspace/bridle)
-- ZEPHYR_TOOLCHAIN_VARIANT: zephyr
-- SDK_VERSION: 0.16.1
-- ZEPHYR_SDK_INSTALL_DIR  : /opt/zephyr-sdk-0.16.1
-- Found Python3: /.../workspace/.env/bin/python (found suitable version "3.8.10", minimum required is "3.8") found components: Interpreter 
-- Cache files will be written to: /home/.../.cache/zephyr
-- Zephyr version: 3.4.99 (/.../workspace/zephyr)
-- Found west (found suitable version "1.1.0", minimum required is "0.14.0")
-- Board: tiac_magpie
CMake Error at zephyr/cmake/modules/extensions.cmake:2764 (message):
  zephyr_set(...  SCOPE snippets) doesn't exists.
Call Stack (most recent call first):
  twister-out/tiac_magpie/tests/drivers/can/timing/drivers.can.timing/zephyr/snippets_generated.cmake:27 (zephyr_set)
  zephyr/cmake/modules/snippets.cmake:94 (include)
  zephyr/cmake/modules/snippets.cmake:121 (zephyr_process_snippets)
  zephyr/cmake/modules/zephyr_default.cmake:129 (include)
  zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:75 (include)
  zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:92 (include_boilerplate)
  zephyr/cmake/package_helper.cmake:115 (find_package)
```

–> **zephyr_set(...  SCOPE snippets) doesn't exists.**

Related to https://github.com/zephyrproject-rtos/zephyr/pull/63232#issuecomment-1739387093

`cat twister-out/tiac_magpie/tests/drivers/can/timing/drivers.can.timing/build.log`:
```
… … …
-- Board: tiac_magpie
CMake Error at zephyr/cmake/modules/extensions.cmake:2764 (message):
  zephyr_set(...  SCOPE snippets) doesn't exists.
Call Stack (most recent call first):
  twister-out/tiac_magpie/tests/drivers/can/timing/drivers.can.timing/zephyr/snippets_generated.cmake:27 (zephyr_set)
… … …
```

`cat twister-out/tiac_magpie/tests/drivers/can/timing/drivers.can.timing/zephyr/snippets_generated.cmake`:
```
# WARNING. THIS FILE IS AUTO-GENERATED. DO NOT MODIFY!
#
# This file contains build system settings derived from your snippets.
# Its contents are an implementation detail that should not be used outside
# of Zephyr's snippets CMake module.
#
# See the Snippets guide in the Zephyr documentation for more information.

###############################################################################
# Global information about all snippets.

# The name of every snippet that was discovered.
set(SNIPPET_NAMES "can-timing-adj" "cdc-acm-console" "usb-console" "xen_dom0")
# The paths to all the snippet.yml files. One snippet
# can have multiple snippet.yml files.
set(SNIPPET_PATHS "/media/slz/lpndevel/devel/TIAC/workspace/bridle/snippets/can-timing-adj/snippet.yml" "/media/slz/lpndevel/devel/TIAC/workspace/bridle/snippets/usb-console/snippet.yml" "/media/slz/lpndevel/devel/TIAC/workspace/zephyr/snippets/cdc-acm-console/snippet.yml" "/media/slz/lpndevel/devel/TIAC/workspace/zephyr/snippets/xen_dom0/snippet.yml")

# Create variable scope for snippets build variables
zephyr_create_scope(snippets)

###############################################################################
# Snippet 'can-timing-adj'

# Common variable appends.
# Appends for board 'tiac_magpie'
if("${BOARD}" STREQUAL "tiac_magpie")
  zephyr_set(EXTRA_DTC_OVERLAY_FILE "/media/slz/lpndevel/devel/TIAC/workspace/bridle/snippets/can-timing-adj/boards/tiac_magpie.overlay" SCOPE snippets APPEND)
endif()
```

It seems `zephyr_create_scope(snippets)` has not the expected effect in case of `west twister …`?

Here a last cross check that shows the unexpected different behaviour between `west build …` and `west twister …`. This check was done without our 3rd party downstream project – Zephyr main branch only (ddd2b490b006e36c2565ac5d53eda272a7a68488).

Also we do not try to use `--extra-args SNIPPET="<valid-snippet-name>"`, but the official recommended way with the test case spec `required_snippets` was used. Thus, to reproduce the CMake bug with original UART API test on board `nucleo_f767zi` we had add the following to lines locally:
```
diff --git a/tests/drivers/uart/uart_async_api/testcase.yaml b/tests/drivers/uart/uart_async_api/testcase.yaml
index e9914d48d9..8b26a7db84 100644
--- a/tests/drivers/uart/uart_async_api/testcase.yaml
+++ b/tests/drivers/uart/uart_async_api/testcase.yaml
@@ -13,6 +13,8 @@ tests:
     harness_config:
       fixture: gpio_loopback
     depends_on: gpio
+    required_snippets:
+      - cdc-acm-console
   drivers.uart.wide:
     filter: CONFIG_UART_CONSOLE and CONFIG_SERIAL_SUPPORT_ASYNC and not CONFIG_UART_MCUX_LPUART
     harness: ztest
```

To reproduce the CMake bug, it's enough to run with `--cmake-only` in both cases, `west build …` and `west twister …`.

### The good case `west build …`

```
west build --cmake-only -p always \
           -b nucleo_f767zi -S cdc-acm-console \
           zephyr/tests/drivers/uart/uart_async_api
```

results in:
```
-- west build: generating a build system
… … …
-- Board: nucleo_f767zi
-- Snippet(s): cdc-acm-console
… … …
-- Found devicetree overlay: /.../workspace/zephyr/snippets/cdc-acm-console/cdc-acm-console.overlay
… … …
Merged configuration '/.../workspace/zephyr/snippets/cdc-acm-console/cdc-acm-console.conf'
… … …
-- Build files have been written to: /.../workspace/build
```

**ALL FINE**

### The buggy path `west twister …`

Note, the patch (above) was applied:
```
west twister --verbose --inline-logs \
             --cmake-only --platform nucleo_f767zi \
             --test tests/drivers/uart/uart_async_api/drivers.uart.async_api
```

results in:
```
INFO    - Using Ninja..
INFO    - Zephyr version: zephyr-v3.4.0-4625-gddd2b490b006
… … …
ERROR   - Cmake build failure: /.../workspace/zephyr/tests/drivers/uart/uart_async_api for nucleo_f767zi
INFO    - 1/1 nucleo_f767zi             tests/drivers/uart/uart_async_api/drivers.uart.async_api  ERROR Cmake build failure (cmake)
INFO    - /.../workspace/twister-out/nucleo_f767zi/tests/drivers/uart/uart_async_api/drivers.uart.async_api/build.log
ERROR   - Loading Zephyr module(s) (Zephyr base): zephyr_default:kconfig
… … …
-- Board: nucleo_f767zi
CMake Error at zephyr/cmake/modules/extensions.cmake:2764 (message):
  zephyr_set(...  SCOPE snippets) doesn't exists.
… … …
```

**NOT EXPECTED**

--- a/cmake/package_helper.cmake
+++ b/cmake/package_helper.cmake
@@ -111,5 +111,16 @@ if(NOT DEFINED MODULES)
   )
 endif()
 
+# Loading Zephyr CMake extension commands, which allows us to overload Zephyr
+# scoping rules.
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE} COMPONENTS extensions)
+
+# Zephyr scoping creates custom targets for handling of properties.
+# However, custom targets cannot be used in CMake script mode.
+# Therefore disable zephyr_set(... SCOPE ...) in package helper as it is not needed.
+function(zephyr_set variable)
+  # This silence the error: zephyr_set(...  SCOPE <scope>) doesn't exists.
+endfunction()
+
 string(REPLACE ";" "," MODULES "${MODULES}")
 find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE} COMPONENTS zephyr_default:${MODULES})

